### PR TITLE
Rearrange the handling of constant slices.

### PR DESCRIPTION
**Pull request by sauclovian-g**
The idea is to split the slice into a reference and a separate static allocation for the body the reference points to. This then allows us to treat the references as normal references to memory, which in turn avoids problems in SAW that arise when it goes to try to enforce disjointness.

Requires accompanying changes in mir-json that produce the new output we read. (These are in [mir-json #73](https://github.com/GaloisInc/mir-json/pull/73).)

See [saw-script #2064](https://github.com/GaloisInc/saw-script/issues/2064) for background.

### CODE REVIEW

**Comment by sauclovian-g** [comment]
Note that this shouldn't get merged without a commit that bumps the mir-json submodule reference, but that should wait until the mir-json changes are ready to merge.

**Comment by sauclovian-g** [comment]
oh, of course none of the tests are going to work without the new mir-json...

File: crucible-mir/src/Mir/Trans.hs
Comment (RyanGlScott) [inline crucible-mir/src/Mir/Trans.hs:397]: Perhaps rename this to something like "`staticSlicePlace`" to reflect its intended purpose.

File: crucible-mir/src/Mir/Trans.hs
Comment (RyanGlScott) [inline crucible-mir/src/Mir/Trans.hs:177]: Yes, slices are encoded as structs with very particular field types, so it's perfectly fine to use `mkSlice` in place of `mkStruct`.

File: crucible-mir/src/Mir/Trans.hs
Comment (RyanGlScott) [inline crucible-mir/src/Mir/Trans.hs:411]: I don't think these comments are relevant anymore, since there is now a suitable `Atom` to use for the `SliceMeta` length. Delete these comments?

File: crucible-mir/src/Mir/Trans.hs
Comment (RyanGlScott) [inline crucible-mir/src/Mir/Trans.hs:162]: I think it's fine to include a historical note about the previous approach used to translate constant slices, but I worry that including the full code (verbatim) will be a bit much if you don't have the full context for what all of these variables are. (For instance, `vec` no longer appears in the new version of the code—I think you'd be better off going through the `git` history if you want the previous implementation.)

My personal preference would be to:

* Describe the steps needed to turn the looked-up `DefId` into a `MirExp` (as you currently do in the bulleted list below).
* Note that MIR slice references used to be translated in a way such that they would be turned into references to constant vectors, but this prevented them from being eligible for equality or overlap checks (citing https://github.com/GaloisInc/saw-script/issues/2064). The current approach instead translates slice references into references to separate global allocations (citing the `ConstStrBody` body case below).

File: crucible-mir/src/Mir/Trans.hs
Comment (RyanGlScott) [inline crucible-mir/src/Mir/Trans.hs:199]: Similarly, I'd favor not including the full code from the previous implementation here. I think it suffices to say that this case it exactly like the `ConstSliceRef` above, but specialized to strings.

**Comment by RyanGlScott** [comment]
> oh, of course none of the tests are going to work without the new mir-json...

Indeed, I think it would be worth bumping the `mir-json` submodule commit to the current state of https://github.com/GaloisInc/mir-json/pull/73 so that we can verify that everything works. When https://github.com/GaloisInc/mir-json/pull/73 lands, we can bump the submodule to the merge commit on `mir-json`'s `master` branch.

Relatedly, I think it would be worth noting somewhere in the commit message that this depends on the changes from https://github.com/GaloisInc/mir-json/pull/73.

File: crucible-mir/src/Mir/Trans.hs
Comment (spernsteiner) [inline crucible-mir/src/Mir/Trans.hs:400]: Would it be possible / reasonable for this to call `staticPlace` instead of duplicating the lookup code?  It looks like the `MirPlace` returned by `staticPlace` contains all the pieces you need

File: crucible-mir/src/Mir/Generator.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/Generator.hs]: Hmm indeed. Someone must have done it that way on purpose then. Wonder why... looks like the contents have changed since so it's probably no longer relevant.

Updated in d4bd0cd0ff03b11e463ef51b795d004c8eaa802e

File: crucible-mir/src/Mir/PP.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/PP.hs]: There was some reason I did that rather than just ignore the value, but I can't remember what and I agree the resulting output is nonsense. I think, since this is for prettyprinting rather than ast-dumping, that it should just skip the length. Updated in 3444637b455edf15c14963403aa69cd2f4192f5e

File: crucible-mir/src/Mir/PP.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/PP.hs]: Obsolete! :-)

File: crucible-mir/src/Mir/Trans.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/Trans.hs]: It got that way because originally it was different from ```staticPlace``` specifically by handling a length instead of a single element... but it's not that simple any more. Updated in 3444637b455edf15c14963403aa69cd2f4192f5e

File: crucible-mir/src/Mir/Trans.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/Trans.hs:180]: Yes, the question was whether it's exactly the same or if there was something special/different about the mkStruct call such that it wasn't using mkSlice for a reason. Which there apparently wasn't. But it wasn't obvious without expanding/looking up a bunch of things.

File: crucible-mir/src/Mir/Trans.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/Trans.hs]: They are not. Removed in 3444637b455edf15c14963403aa69cd2f4192f5e

File: crucible-mir/src/Mir/Trans.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/Trans.hs]: That comment's there as a semi-proof to myself that I haven't accidentally changed the behavior.

There's a lot of things here where you can write the same action as either ```a``` or ```A``` (so to speak) and then assorted combinations of ```Ab```, ```Bc```, ```bC```, etc. such that if you started with ```((Ab,cD),ef)``` and want to change it to ```(a, (bCd), EF)``` or similar you want to be pretty careful that what you've changed it to is actually equivalent, and neither missing a step nor replacing ```c``` with ```C'``` instead of ```C``` when you didn't mean to. The first couple times I got it wrong (even laying it out explicitly), hence the high level of paranoia.

Now that it's running maybe it should just be flushed, though.

Let me chew on it over the weekend.

File: crucible-mir/src/Mir/Trans.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/Trans.hs]: This one's now gone entirely anyway.

File: crucible-mir/src/Mir/Trans.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/Trans.hs:384]: It could, but it seems the wrong way to abstract things (these are two different ways to look up a static and construct a Place for it, and having the more complex one call the simpler one and then munge the results doesn't seem right)

So... I'd rather not I think

**Comment by sauclovian-g** [comment]
> I think it would be worth bumping the `mir-json` submodule commit to the current state of [GaloisInc/mir-json#73](https://github.com/GaloisInc/mir-json/pull/73) so that we can verify that everything works. When [GaloisInc/mir-json#73](https://github.com/GaloisInc/mir-json/pull/73) lands, we can bump the submodule to the merge commit on `mir-json`'s `master` branch.

I have done this.

> Relatedly, I think it would be worth noting somewhere in the commit message that this depends on the changes from [GaloisInc/mir-json#73](https://github.com/GaloisInc/mir-json/pull/73).

Now each of the descriptions references the other one.

File: crucible-mir/src/Mir/Trans.hs
Comment (spernsteiner) [reply crucible-mir/src/Mir/Trans.hs:384]: FWIW I think one interpretation of `staticSlicePlace` is that it obtains a reference to the array `did` (e.g. `&[u8; 10]`), which is exactly the output of `staticPlace`, and then converts that array reference to a slice reference (`&[u8]`).  But implementation-wise I think it's fine either way.

File: crucible-mir/src/Mir/Mir.hs
Comment (spernsteiner) [reply crucible-mir/src/Mir/Mir.hs:499]: > `ConstSliceRef` is a reference to a static allocation for a slice (of type `[T]`) or a string (or type `str`).

To be clear, the type of `ConstSliceRef` is `&[T]` or `&str`

> In the former case, the allocation is backed by a constant array (i.e., a `ConstArray`), and in the latter case, the allocation is backed by a constant string (i.e., a `ConstStrBody`).

IMO there's no need to mandate this if we view `ConstStrBody` as shorthand for a `ConstArray` of bytes.  There's no issue with allowing `ConstStrBody` for `&[u8]` or `&[u8; N]`.

Comment (RyanGlScott) [reply crucible-mir/src/Mir/Mir.hs:499]: > To be clear, the type of `ConstSliceRef` is `&[T]` or `&str`

Oh, certainly! I didn't mean to imply that the overall type of the reference was `[T]` or `str`. I was more describing how what type of constant the static allocation would be creating internally.

> IMO there's no need to mandate this if we view `ConstStrBody` as shorthand for a `ConstArray` of bytes. There's no issue with allowing `ConstStrBody` for `&[u8]` or `&[u8; N]`.

That's fair. That being said, in practice we only use `ConstStrBody` as part of creating static allocations for `&str`s, correct? It's possible that we might change this in the future, but I think it's helpful to make a note of under what circumstances `ConstStrBody` can arise today. (It wasn't obvious to me at a first glance.)

File: crucible-mir/src/Mir/Trans.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/Trans.hs:384]: That is also true. hrm...

I guess at this point the path of least resistance is to not make more changes...

File: crucible-mir/src/Mir/Trans.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/Trans.hs]: Conclusion: it should just be flushed. It is only of interest if the new code turns out not to work in some corner case, and in that case fetching things out of history is fine.

File: crucible-mir/src/Mir/Mir.hs
Comment (sauclovian-g) [reply crucible-mir/src/Mir/Mir.hs:501]: That is correct. I've added some notes.

**Review by RyanGlScott — APPROVED** [review]

**Comment by sauclovian-g** [comment]
I force-pushed to squash the reference to the pre-merged version in mir-json. Perhaps not strictly necessary (since that version still exists) but it's untidy to leave it lying around.

**Pull request merged**
